### PR TITLE
[MAINTENANCE] convert docs build scripts from bash to python

### PR DESCRIPTION
**Pull request by tyler-hoffman**
This migrates our docs build shell scripts to python. This flow was previously a mix of python and shell scripts

**Overview**
The general changes are:
* The `build_docs` shell script is to only be used by netflify, and it is a thin wrapper around `invoke docs`
* `invoke docs` calls into a python class, rather than running a shell script. Previously, we would run one of 2 bash scripts. These have both been converted to public methods in the new DocsBuilder class.
* A best effort was done to abstract logic into methods, without any behavioral changes.

**Things intentionally kept out of this PR**
* any functional changes
* merging the work from `docs/prepare_prior_versions.py` into this
* unit testing - this would be a largish effort for an internal tool, and we already have netlify running the build in CI to catch big issues

- [x] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [x] Code is linted - run `invoke lint` (uses `black` + `ruff`)
- [x] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: docs/logging.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class Logger:
```
Comment (tyler-hoffman) [inline docs/logging.py:1]: Do we really care about this? I kinda like the headers and coloring, but maybe we should just use the build in `Logger`?

File: tasks.py [hunk 2]
```diff
@@ -278,7 +279,7 @@ def type_check(  # noqa: PLR0913, PLR0912
         print(f"  Clearing {mypy_cache} ... ", end="")
         try:
             shutil.rmtree(mypy_cache)
-            print("✅"),
+            print("✅")
```
Comment (tyler-hoffman) [inline tasks.py:282]: scope creep: this was a typo before.

File: tasks.py [hunk 3]
```diff
@@ -666,9 +667,21 @@ def docs(
     )
 
     print("Running invoke docs from:", repo_root)
-    old_pwd = pathlib.Path.cwd()
+    old_cwd = pathlib.Path.cwd()
```
Comment (tyler-hoffman) [inline tasks.py:670]: looks like a typo before. pwd = print working directory, cwd = current working directory.

File: docs/logging.py (hunk 1 above)
Comment (billdirks) [reply docs/logging.py:1]: Since this is already here I think it's fine to use it.

If we ant to use colors, here's a stack-overflow answer that shows how to configure a built in logger formatter so you can log different levels with different colors.

File: docs/logging.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+        LINE = "================================================================================"
+        Logger.print(LINE)
+        Logger.print(string)
+        Logger.print(LINE)
+
+    @staticmethod
+    def print(string: str) -> None:
+        ORANGE = "\033[38;5;208m"
+        END = "\033[1;37;0m"
+        print(ORANGE + string + END)
```
Comment (billdirks) [inline docs/logging.py:15]: You use Logger.print for `print_header` but `print` here. Maybe use `Logger.print` here too?

File: reqs/requirements-dev-contrib.txt [hunk 5]
```diff
@@ -3,5 +3,6 @@ black[jupyter]==23.10.1
 invoke>=2.0.0
 mypy==1.7.1
 pre-commit>=2.21.0
+requests>=2.20
```
Comment (billdirks) [inline reqs/requirements-dev-contrib.txt]: We have this requirement at the top level `requirements.txt` file. I'm not sure if we also need it here.

File: tasks.py [hunk 6]
```diff
@@ -666,9 +667,21 @@ def docs(
...
     docusaurus_dir = repo_root / "docs/docusaurus"
     os.chdir(docusaurus_dir)
+    pull_request = os.environ.get(  # noqa: TID251 # os.enrivon allowed in config files
```
Comment (billdirks) [inline tasks.py]: nit: `s/enrivon/environ/`

Comment (billdirks) [reply tasks.py]: Could you add a comment that netlify will set the env variable?

File: docs/docs_build.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    @contextmanager
+    def _load_zip(self, url: str) -> Generator[zipfile.ZipFile, None, None]:
+        response = requests.get(url)
+        zip_data = BytesIO(response.content)
+        with zipfile.ZipFile(zip_data, "r") as zip_ref:
+            yield zip_ref
+
+    def _load_files(self) -> None:
+        """Load oss_docs_versions zip and relevant versions from github.
+        oss_docs_versions contains the versioned docs to be used later by prepare_prior_versions, as well
```
Comment (billdirks) [inline docs/docs_build.py:82]: Add a blank space between initial summary line of docstring and rest of docstring: https://google.github.io/styleguide/pyguide.html#381-docstrings

File: docs/docs_build.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+                    / f"versioned_code/great_expectations-{version}"
+                )
+                new_location = (
+                    self._current_directory / f"versioned_code/version-{version}"
+                )
+                shutil.move(str(old_location), str(new_location))
+
+    def _invoke_api_docs(self) -> None:
+        """Invokes the invoke api-docs command.
+        If this is a non-PR running on netflify, we use the latest tag. Otherwise, we use the current branch.
```
Comment (billdirks) [inline docs/docs_build.py]: nit: `s/netflify/netlify/` 😛

**Review by billdirks — APPROVED** [review]
LGTM. Nice to see it all in python! I left a few minor comments.

File: reqs/requirements-dev-contrib.txt (hunk 5 above)
Comment (tyler-hoffman) [reply reqs/requirements-dev-contrib.txt]: I didn't think we would either, but then I failed CI without it. https://github.com/great-expectations/great_expectations/actions/runs/7493298388/job/20398915090. Alright. I take that back somewhat. Something's up. Investigating now

File: tasks.py (hunk 6 above)
Comment (tyler-hoffman) [reply tasks.py]: yup, will do!

File: docs/logging.py (hunk 4 above)
Comment (tyler-hoffman) [reply docs/logging.py:15]: We're in the method `Logger.print` 😄
